docs/source/torch.compiler_dynamic_shapes.md

```diff
@@ -1,129 +1,474 @@
+f(torch.rand(40))
+f(torch.rand(50))
+```
+
+In the code above, we specialize that the graph requires an input size of 10, in which
+case it will return `x * 10`. If the input size is less than 30, it will return `x * 200`.
+In the output, you can see that this creates three graphs.
+
+See the corresponding <a href="/_static/img/dynamic_shapes/tlparse3_specialization.png" target="_blank">tlparse output</a>
+
 
-See also: [The dynamic shapes manual](https://docs.google.com/document/d/1GgvOe7C8_NVOMLOCwDaYV1mXXyHMXY7ExoewHqooxrs/edit#heading=h.fh8zzonyw8ng)
+This is how graphs created for the above function:
 
-## Motivation
+```{image} _static/img/dynamic_shapes/dynamic_shapes_example_specialization.png
+```
+
+(enable-dynamic-behavior)=
+## Enabling Dynamic Behavior
+
+There are the following ways to make things dynamic:
 
-Deep learning compilers commonly only work for static shapes, that is to say, they produced compiled programs which only work for a single specific configuration of input shapes, and must recompile if any input shape changes. This assumption works great for the majority of commonly run deep learning models today, but there are a few situations where it is insufficient:
+* Automatic dynamic
+* torch.compile (dynamic=true)
+* Profile-Guided Optimization (PGO)
+* Compiler Collective
+* Users annotations
 
-- Some dimensions, such as batch size or sequence length, may vary. For example, an inference service performing adaptive batching will execute inference requests with varying batch sizes depending on how many requests it received within its batching window. We may also want to consider padding out variable size sequences only to the maximum sequence length within a batch, which may vary from batch-to-batch.
-- Some models exhibit data-dependent output shapes, that is to say, the size of their outputs and intermediates may depend on the actual input data which may vary across runs. For example, detection models may first generate a variable number of potential bounding boxes before running a more expensive image recognition model to identify if the subject is in a bounding box. The number of bounding boxes is data dependent.
-- One particularly important case of data-dependent shapes occurs when dealing with sparse representations, such as sparse tensors, jagged tensors, and graph neural networks. In all of these cases, the amount of data to be processed depends on the sparse structure of the problem, which will typically vary in a data-dependent way.
+Read below about each of this options.
 
-In supporting dynamic shapes, we chose not to support dynamic rank programs, e.g., programs whose inputs tensors change in dimensionality, as this pattern rarely occurs in real-world deep learning programs, and it avoids the need to reason inductively over symbolic lists of shapes.
+### Automatic dynamic
 
-## Abridged public API
+This is the default behavior where Dynamo makes an input dynamic if it
+sees different values for it.
```

Comment:
I think we should expand on this more. It should say something like this

By default, torch.compile performs the initial compilation assuming static shapes. However, it tracks the input sizes used during that first compilation. If a recompile is triggered, it uses this information to identify which dimensions have changed and marks those as dynamic for the second compilation. This mechanism is known as automatic dynamic, and it’s the default behavior in PT2